### PR TITLE
#4621 Temporary Display Info slice 2

### LINKED ISSUE #4621: Temporary Popup Slice 2: add the ability to view the panel using the snap of the last run
\[This issue was auto\-generated using ProductBoard\.\]

<p></p>

**Draft specification:** [**https:\/\/www\.notion\.so\/pixiebrix\/Ephemeral\-Temporary\-Document\-Renderer\-acd663a90fb64444ae372404ea6bda8a**](https://www.notion.so/pixiebrix/Ephemeral-Temporary-Document-Renderer-acd663a90fb64444ae372404ea6bda8a)

<p></p>

- **Slice 2: add the ability to view the panel using the snap of the last run**

    - In the Preview pane of the data panel, add a “Preview from Snapshot” button

    - On click, send a message to the content script with the 1\) brick definition, and 2\) latest snapshot of data\. \(See existing preview code\)

  <p></p>

### PR DESCRIPTION
**Pull request by BLoe**
## What does this PR do?

- Closes #4621 

## Reviewer Tips

- The new hook is entirely duplicate data aggregation code from `BlockPreview`, but then it calls `runRendererBlock` instead of `runBlock`

## Discussion

- Having to pass the `title`, `extensionId`, and `runId` into the `runRendererBlock` function is messy and will have to be re-worked if/when we expand this functionality beyond just temp display bricks

## Demo

https://www.loom.com/share/1be9c486caec45ad93275a419667e64b

## Checklist

- [x] Add tests
- [ ] Run Storybook and manually confirm that all stories are working
- [x] Designate a primary reviewer

### CODE REVIEW

File: src/components/documentBuilder/preview/DocumentPreview.test.tsx [hunk 1]
```diff
@@ -98,3 +128,68 @@ describe("Add new element", () => {
...
   });
 });
+
+describe("Show live preview", () => {
+  beforeAll(() => {
+    jest.useFakeTimers();
+  });
+
+  afterAll(() => {
+    jest.runAllTimers();
```
Comment (BALEHOK) [inline src/components/documentBuilder/preview/DocumentPreview.test.tsx:138]: You can call this after each test to make a sort of cleanup and have the next test start fresh.

File: src/components/documentBuilder/preview/DocumentPreview.tsx [hunk 2]
```diff
@@ -62,43 +74,73 @@ const DocumentPreview = ({
...
+          >
+            Show Live Preview <FontAwesomeIcon icon={faExternalLinkAlt} />
+          </Button>
+          <br />
+          {previewError && (
+            <Alert variant="danger">{getErrorMessage(previewError)}</Alert>
+          )}
+          <hr />
+        </>
       )}
```
Comment (BALEHOK) [inline src/components/documentBuilder/preview/DocumentPreview.tsx:106]: Isn't it a responsibility of Preview Tab to show different buttons and messages? Maybe it's better to place this logic in `DataPanel` (directly or add a `DocumentPreviewTab`)?
`DocumentPreview` was meant as a preview/edit control for the `DocumentEditor`. It shouldn't know about running a pipeline and runtime errors. `DocumentPreview` lives in components and ideally it stays independent from items from `dataPanel` folder.

**Review by BALEHOK — COMMENTED** [review]
Is it possible to reuse the code from the hook in BlockPreview? Those parts do not look trivial, maintaining two versions of rendering a preview can be problematic.

**Comment by BLoe** [comment]
@BALEHOK There is definitely a lot of tech debt in this implementation. I absolutely agree that separating concerns from the DocumentPreview component would be a cleaner architecture. Unfortunately, this feature is blocking some things around marketing and our client TaskUs, so we're trying to get out a working-but-really-hacky version today.

I am going to write up a new github ticket now with considerations from my work on this PR. It'll cover improvements I think we can make to the way blocks are executed outside of normal extension runs, as well as cleaning up this live preview functionality so that it can apply to any instance of the document builder. I'll link it here when I finish writing it up.

**Review by twschiller — APPROVED** [review]

**Pull request merged**